internal/consumer/dlq.go

```diff
@@ -109,4 +148,197 @@ func NewDLQNoop() DLQ {
 	return &dlqNoop{}
 }
 func (d *dlqNoop) Add(m kafka.Message) error { return nil }
-func (d *dlqNoop) Close()                    {}
+func (d *dlqNoop) Stop()                     {}
+func (d *dlqNoop) Start() error              { return nil }
+
+func newBufferedDLQImpl(topic string, producer sarama.SyncProducer, timeLimit <-chan time.Time, sizeLimit <-chan time.Time, tally tally.Scope, log *zap.Logger) *bufferedDLQImpl {
+	return &bufferedDLQImpl{
+		topic:            topic,
+		producer:         producer,
+		timeLimit:        timeLimit,
+		sizeLimit:        sizeLimit,
+		stopC:            make(chan struct{}),
+		closedC:          make(chan struct{}),
+		messageBufferMap: newMessageBufferMap(),
+		errorMap:         newErrorMap(),
+		scope:            tally,
+		logger:           log,
+		lifecycle:        util.NewRunLifecycle(fmt.Sprintf("dlq-%s", topic), log),
+	}
+}
+
+func (d *bufferedDLQImpl) Start() error {
+	return d.lifecycle.Start(func() error {
+		go d.eventLoop()
+		return nil
+	})
+}
+
+func (d *bufferedDLQImpl) Stop() {
+	d.lifecycle.Stop(func() {
+		close(d.stopC)
+		<-d.closedC
+	})
+}
+
+func (d *bufferedDLQImpl) Add(msg kafka.Message) error {
+	key := kafkaMessageKey(msg)
+	sm := d.newSaramaMessage(msg)
+	d.messageBufferMap.Add(key, sm)
+	return d.errorMap.GetAndRemove(key) // errorMap.Get blocks until there is error or nil for this message
+}
+
+func (d *bufferedDLQImpl) eventLoop() {
```

Comment:
This code can be simplified quite a bit, if you use the sarama.AsyncProducer which already has most of this logic in place. Are there things that the AsyncProducer doesn't support that you need here ? The AsyncProducer can do the following:

* Buffer and batch based on time / size
* Block the sender when the inflight > threshold 
* Response channel that sends success / failure for every message attempted to publish

If you go with the AsyncProducer all you need is an inFlightMap[msgID]chan where chan is the response channel the caller (DLQ.Add) is waiting on. And then you need a goroutine that 
* dequeues messages from sarama.AsyncProducer.Successes()  / sarama.AsyncProducer.Errors() 
* looks up the inFlightMap to find response (chan error) for msgID 
* Sets error to nil or error

Reply:
Thats a great idea. In fact, with `sarama.ProducerMessage.Metadata`, we dont even need the `inFlightMap`.